### PR TITLE
mono: Fix compile/runtime error on El Cap Beta 3

### PR DESCRIPTION
**Pull request by felixbuenemann**
This fixes the error "could not discover the mach TLS offset" which
prohibits both running and compiling mono on OS X 10.11 (15A216g).

The patch was backported from current mono/mono master. Because the
problem also affects binaries compiled before upgrading to El Capitan
the patch is *not* scoped to `MacOS.version >= :el_capitan`.

See [Xamarin Bugzilla Bug 31761](https://bugzilla.xamarin.com/show_bug.cgi?id=31761) and PR mono/mono#1919 for more info.

### CODE REVIEW

**Comment by felixbuenemann** [comment]
Is there a reason that the brew test-bot is requesting access to my github account's associated organizations and it's repos when I click on the Details link?

**Comment by DomT4** [comment]
We have it set at the least amount of permissions we're told the plugins require, I believe. CC @mikemcquaid

**Comment by andrewbranch** [comment]
You, sir, are a life saver.

**Comment by mikemcquaid** [comment]
@felixbuenemann screen shot 2015-07-09 at 23 59 01

The only thing it requests is to see what organisation teams you're a member of (to see if you're in the Homebrew owners team).

**Comment by jonlbauer** [comment]
Hi All,

So I'm a total newbie, but I just encountered this issue with the Public Beta release of El Capitan...

I have the newest mono and the 3.10.0 versions installed, and I get this error and a freeze up on both versions. BTW, I installed 4.0.2.5 from scratch, and the mono installer didn't set up a path for me either. So that's broken too...

Can someone please tell me, in plain English :-) how I can fix it?

- Thanx in advance!
- Jon

**Comment by felixbuenemann** [comment]
@mikemcquaid Thanks for the clarification. I thought it was requesting access to organization repos.

**Comment by felixbuenemann** [comment]
@jonlbauer The compile will fail if there is an unfixed version of mono in your path. In the normal update case the old homebrew version is unlinked before the compile is started, so that's not an issue. But if you have multiple mono versions installed that might be the problem.

**Comment by jonlbauer** [comment]
Hm. I don't know how to do a compile on my own... Can you summarize the process for me? I"m guessing that I need to download the fix that is summarized here, merge it into the version that I have, and then re-compile it?

- Jon

**Comment by jonlbauer** [comment]
Oh wait a sec - are you saying that I should uninstall 3.10.0 and then try the normal install method? But what about the fix that is being outlined here? I don't need to worry about that?

- Jon

**Comment by felixbuenemann** [comment]
@jonlbauer I think you can install from a url, try `brew install https://raw.githubusercontent.com/felixbuenemann/homebrew/fix-mono-el-cap-tls-offset/Library/Formula/mono.rb`

**Comment by jonlbauer** [comment]
That didn't work.

If I install home-brew, and then install mono through it, will it be using the fix that's been outlined here?

- Jon

File: Library/Formula/mono.rb [hunk 1]
```diff
@@ -14,6 +15,10 @@ class Mono < Formula
     sha256 "80bddaefcd9c136bc763797cff98f3aa07ec758e348037fd24f07d2b1b795fbb" => :mountain_lion
   end
 
+  # Fix compile and runtime error on OS X 10.11 Beta 3
+  # See: https://github.com/mono/mono/pull/1919
+  patch :DATA
```
Comment (mikemcquaid) [inline Library/Formula/mono.rb]: Please use `https://github.com/mono/mono/pull/1919` as the patch URL, thanks!

File: Library/Formula/mono.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@ class Mono < Formula
   homepage "http://www.mono-project.com/"
   url "http://download.mono-project.com/sources/mono/mono-4.0.2.5.tar.bz2"
   sha256 "b074584eea5bbaaf29362486a69d70abe53d0d2feb334f231fa9c841cf6fd651"
+  revision 1
```
Comment (mikemcquaid) [inline Library/Formula/mono.rb]: Don't think this needs a `revision` given it wasn't installing on 10.11 before.

Comment (felixbuenemann) [inline Library/Formula/mono.rb]: I bumped the revision because mono compiled eg. on yosemite will fail on el cap unless this fix is included. That's also the reason the patch is replied regardless of OS X version.

File: Library/Formula/mono.rb (hunk 1 above)
Comment (felixbuenemann) [inline Library/Formula/mono.rb]: I didn't think of that. I'll push a new rev, thanks!

**Comment by felixbuenemann** [comment]
@jonlbauer You could install homebrew, uninstall your current mono and install using the command posted above or using `brew install mono` as soon as this pull request is merged. Please take some time to get yourself aquainted with how homebrew works, a pull request is for developer discussions not user support and using it as such causes a lot of noise for the devs involved. Thanks!

File: Library/Formula/mono.rb (hunk 2 above)
Comment (mikemcquaid) [inline Library/Formula/mono.rb]: I don't think it's worth forcing an upgrade on everyone for that edge-case, thanks.

Comment (felixbuenemann) [inline Library/Formula/mono.rb]: OK, I'll drop the revision.

**Comment by felixbuenemann** [comment]
Pushed a new commit to address @mikemcquaid's feedback.

**Comment by mikemcquaid** [comment]
Thanks @felixbuenemann!

**Comment by bosmacs** [comment]
FYI, you may also want to include the x86 patch for the same problem: 
~~~rb
  # and for x86
  patch do
    url "https://github.com/mono/mono/pull/1924.diff"
    sha256 "62e9d2ad134558c3fcf9c75e260c6091e390b23b4f320e79184136b76c021a39"
  end
~~~

**Comment by felixbuenemann** [comment]
@bosmacs Is there a case where mono would compile as x86 on El Cap?

**Comment by bosmacs** [comment]
I thought it might affect code generation for 32-bit assemblies, but in brief testing that appears not to be the case. So perhaps it’s not needed after all.
